File: Robotics Kinematics and Dynamics Course/Chapter12/Lesson2/Chapter12_Lesson2.m
function [f_list, n_list, tau] = ne_backward(F_list, N_list, R_ip1_i_list, ...
                                             p_i_list, z_list, joint_type_list, ...
                                             f_tip, n_tip)
% F_list, N_list: 3xN
% R_ip1_i_list: 3x3xN
% p_i_list, z_list: 3xN
% joint_type_list: 1xN char array: 'R' or 'P'
N = size(F_list, 2);
f_list = zeros(3, N);
n_list = zeros(3, N);
tau = zeros(1, N);

f_next = f_tip(:);
n_next = n_tip(:);

for i = N:-1:1
    R_ip1_i = R_ip1_i_list(:, :, i);
    p_i = p_i_list(:, i);
    F_i = F_list(:, i);
    N_i = N_list(:, i);
    z_i = z_list(:, i);
    jt = joint_type_list(i);

    f_child_i = R_ip1_i * f_next;
    n_child_i = R_ip1_i * n_next + cross(p_i, f_child_i);

    f_i = F_i + f_child_i;
    n_i = N_i + n_child_i;

    f_list(:, i) = f_i;
    n_list(:, i) = n_i;

    if jt == 'R'
        tau(i) = z_i.' * n_i;
    else
        tau(i) = z_i.' * f_i;
    end

    f_next = f_i;
    n_next = n_i;
end
end
      

File: Tutorials/Codes/Robotics_Kinematics_and_Dynamics_Course/Chapter12/Lesson2/Chapter12_Lesson2.cpp
#include <iostream>
#include <vector>
#include <Eigen/Dense>

struct LinkDynamics {
    Eigen::Vector3d F;   // inertial force
    Eigen::Vector3d N;   // inertial moment
    Eigen::Matrix3d R_ip1_i; // ^i R_(i+1)
    Eigen::Vector3d p_i; // from frame i to i+1 in frame i
    Eigen::Vector3d z;   // joint axis
    char joint_type;     // 'R' or 'P'
};

void newtonEulerBackward(const std::vector<LinkDynamics>& links,
                         const Eigen::Vector3d& f_tip,
                         const Eigen::Vector3d& n_tip,
                         std::vector<Eigen::Vector3d>& f_list,
                         std::vector<Eigen::Vector3d>& n_list,
                         std::vector<double>& tau_list)
{
    int n = static_cast<int>(links.size());
    f_list.assign(n, Eigen::Vector3d::Zero());
    n_list.assign(n, Eigen::Vector3d::Zero());
    tau_list.assign(n, 0.0);

    Eigen::Vector3d f_next = f_tip;
    Eigen::Vector3d n_next = n_tip;

    for (int i = n - 1; i >= 0; --i) {
        const auto& L = links[i];

        Eigen::Vector3d f_child_i = L.R_ip1_i * f_next;
        Eigen::Vector3d n_child_i = L.R_ip1_i * n_next
                                  + L.p_i.cross(f_child_i);

        Eigen::Vector3d f_i = L.F + f_child_i;
        Eigen::Vector3d n_i = L.N + n_child_i;

        f_list[i] = f_i;
        n_list[i] = n_i;

        double tau_i;
        if (L.joint_type == 'R') {
            tau_i = L.z.dot(n_i);
        } else { // 'P'
            tau_i = L.z.dot(f_i);
        }
        tau_list[i] = tau_i;

        f_next = f_i;
        n_next = n_i;
    }
}

int main() {
    std::vector<LinkDynamics> links(2);
    // Initialize with some toy data (in a real application, fill from model)
    links[0].F = Eigen::Vector3d(1.0, 0.0, 0.0);
    links[0].N = Eigen::Vector3d(0.0, 0.0, 0.2);
    links[0].R_ip1_i = Eigen::Matrix3d::Identity();
    links[0].p_i = Eigen::Vector3d(0.5, 0.0, 0.0);
    links[0].z = Eigen::Vector3d(0.0, 0.0, 1.0);
    links[0].joint_type = 'R';

    links[1].F = Eigen::Vector3d(0.5, 0.0, 0.0);
    links[1].N = Eigen::Vector3d(0.0, 0.0, 0.1);
    links[1].R_ip1_i = Eigen::Matrix3d::Identity();
    links[1].p_i = Eigen::Vector3d(0.3, 0.0, 0.0);
    links[1].z = Eigen::Vector3d(0.0, 0.0, 1.0);
    links[1].joint_type = 'R';

    Eigen::Vector3d f_tip = Eigen::Vector3d::Zero();
    Eigen::Vector3d n_tip = Eigen::Vector3d::Zero();

    std::vector<Eigen::Vector3d> f_list, n_list;
    std::vector<double> tau_list;

    newtonEulerBackward(links, f_tip, n_tip, f_list, n_list, tau_list);

    std::cout << "Joint torques: ";
    for (double tau : tau_list) {
        std::cout << tau << " ";
    }
    std::cout << std::endl;
    return 0;
}
      
